File: turborepo/packages/api-client/interceptors.ts
import { AxiosError, AxiosInstance, InternalAxiosRequestConfig, AxiosResponse } from 'axios'
import { TokenDto } from '@repo/common/model/dto/token-dto'
import { isTokenExpired, isTokenExpiringSoon } from '@repo/common/utils/auth-utils'
import { startLoading, stopLoading } from '@repo/common/utils/loading-bus'

export interface AuthState {
  tokenResponse: TokenDto.TokenResponse | null
  tokenIssuedAt: number | null
  refreshToken: () => Promise<boolean>
  clearTokenResponse: () => void
  isInitialized: () => boolean
  initializeAuth: () => Promise<void>
}

type RetryRequestConfig = InternalAxiosRequestConfig & {
  _retry?: boolean
  showErrorPopup?: boolean
}

type QueueEntry = {
  resolve: (value: string | null) => void
  reject: (error: unknown) => void
}

let isRefreshing = false
let failedQueue: QueueEntry[] = []

const processQueue = (error: unknown, token: string | null = null) => {
  failedQueue.forEach(({ resolve, reject }) => {
    if (error) {
      reject(error)
    } else {
      resolve(token)
    }
  })

  failedQueue = []
}

export const setupAxiosInterceptors = (
  axiosInstance: AxiosInstance,
  getAuthState: () => AuthState,
  showError: (message: string) => void,
  onRedirectToLogin: () => void,
  onForbidden: () => void = () => {},
) => {
  // Request Interceptor - 토큰 자동 첨부
  axiosInstance.interceptors.request.use(
    async (config: InternalAxiosRequestConfig) => {
      if (config.showLoading !== false) startLoading()

      const isPublicRequest = config.url?.includes('/auth/login') ||
                             config.url?.includes('/auth/register') ||
                             config.url?.includes('/auth/refresh')

      if (isPublicRequest) {
        return config
      }

      let { tokenResponse, refreshToken, clearTokenResponse } = getAuthState()
      const { isInitialized, initializeAuth } = getAuthState()

      if (!isInitialized()) {
        await initializeAuth()
        const updatedState = getAuthState()
        tokenResponse = updatedState.tokenResponse
        refreshToken = updatedState.refreshToken
        clearTokenResponse = updatedState.clearTokenResponse
      }

      if (tokenResponse?.accessToken) {
        const isLogoutRequest = config.url?.includes('/auth/logout')
        const { tokenIssuedAt } = getAuthState()

        if (!isLogoutRequest && isTokenExpired(tokenResponse, tokenIssuedAt)) {
          if (config.showLoading !== false) stopLoading()
          clearTokenResponse()
          onRedirectToLogin()
          return Promise.reject(new Error('Token expired'))
        }

        if (!isLogoutRequest && isTokenExpiringSoon(tokenResponse, tokenIssuedAt, 5)) {
          if (!isRefreshing) {
            isRefreshing = true

            try {
              const refreshSuccess = await refreshToken()
              if (refreshSuccess) {
                const newToken = getAuthState().tokenResponse?.accessToken
                processQueue(null, newToken)
                config.headers = config.headers || {}
                config.headers['Authorization'] = `Bearer ${newToken}`
              } else {
                processQueue(new Error('Token refresh failed'), null)
                if (config.showLoading !== false) stopLoading()
                clearTokenResponse()
                onRedirectToLogin()
                return Promise.reject(new Error('Token refresh failed'))
              }
            } catch (error) {
              processQueue(error, null)
              if (config.showLoading !== false) stopLoading()
              clearTokenResponse()
              onRedirectToLogin()
              return Promise.reject(error)
            } finally {
              isRefreshing = false
            }
          } else {
            return new Promise((resolve, reject) => {
              failedQueue.push({ resolve, reject })
            }).then((token) => {
              config.headers = config.headers || {}
              config.headers['Authorization'] = `Bearer ${token}`
              return config
            })
          }
        } else {
          config.headers = config.headers || {}
          config.headers['Authorization'] = `Bearer ${tokenResponse.accessToken}`
        }
      }

      return config
    },
    (error) => {
      return Promise.reject(error)
    }
  )

  // Response Interceptor - 에러 처리
  axiosInstance.interceptors.response.use(
    (response: AxiosResponse) => {
      if (response.config.showLoading !== false) stopLoading()
      return response
    },
    async (error: AxiosError<{ message?: string; error?: string }>) => {
      const originalRequest = error.config as RetryRequestConfig | undefined
      if (error.config?.showLoading !== false) stopLoading()

      if (!originalRequest) {
        const errorMessage = error.response?.data?.message ||
          error.response?.data?.error ||
          error.message ||
          '요청 처리 중 오류가 발생했습니다.'

        showError(errorMessage)
        return Promise.reject(error)
      }

      const isAuthRequest =
        originalRequest.url?.includes('/auth/logout') ||
        originalRequest.url?.includes('/auth/refresh')

      const errorMessage = error.response?.data?.message ||
        error.response?.data?.error ||
        error.message ||
        '요청 처리 중 오류가 발생했습니다.'

      const shouldHandle401Refresh =
        error.response?.status === 401 && !originalRequest._retry && !isAuthRequest

      if (originalRequest.showErrorPopup !== false && !isAuthRequest && !shouldHandle401Refresh) {
        showError(errorMessage)
      }

      if (error.response?.status === 403) {
        onForbidden()
      }

      if (shouldHandle401Refresh) {
        originalRequest._retry = true

        const { tokenResponse, refreshToken, clearTokenResponse } = getAuthState()

        if (tokenResponse && !isRefreshing) {
          isRefreshing = true

          try {
            const refreshSuccess = await refreshToken()

            if (refreshSuccess) {
              const newToken = getAuthState().tokenResponse?.accessToken
              processQueue(null, newToken)

              originalRequest.headers = originalRequest.headers || {}
              originalRequest.headers['Authorization'] = `Bearer ${newToken}`
              return axiosInstance(originalRequest)
            } else {
              processQueue(error, null)
              if (originalRequest.showErrorPopup !== false) {
                showError(errorMessage)
              }
              clearTokenResponse()
              onRedirectToLogin()
            }
          } catch (refreshError) {
            processQueue(refreshError, null)
            if (originalRequest.showErrorPopup !== false) {
              showError(errorMessage)
            }
            clearTokenResponse()
            onRedirectToLogin()
          } finally {
            isRefreshing = false
          }
        } else if (isRefreshing) {
          return new Promise((resolve, reject) => {
            failedQueue.push({ resolve, reject })
          }).then((token) => {
            originalRequest.headers = originalRequest.headers || {}
            originalRequest.headers['Authorization'] = `Bearer ${token}`
            return axiosInstance(originalRequest)
          }).catch((err) => {
            return Promise.reject(err)
          })
        } else {
          if (originalRequest.showErrorPopup !== false) {
            showError(errorMessage)
          }
          clearTokenResponse()
          onRedirectToLogin()
        }
      }

      return Promise.reject(error)
    }
  )
}

File: mobile-monorepo/packages/api_client/lib/src/core/interceptors.dart
import 'dart:async';

import 'package:common/common.dart';
import 'package:dio/dio.dart';

import '../store/auth_token_provider.dart';
import 'auth_event_bus.dart';

const _publicPathFragments = <String>[
  '/auth/login',
  '/app/auth/login',
  '/auth/join',
  '/auth/refresh',
];

const _logoutPathFragment = '/auth/logout';

bool _isPublicRequest(String? path) {
  if (path == null) return false;
  return _publicPathFragments.any(path.contains);
}

bool _isLogoutRequest(String? path) {
  if (path == null) return false;
  return path.contains(_logoutPathFragment);
}

bool _shouldShowLoading(RequestOptions options) {
  return options.extra['showLoading'] != false;
}

bool _shouldShowErrorPopup(RequestOptions options) {
  return options.extra['showErrorPopup'] != false;
}

String _resolveErrorMessage(DioException error) {
  if (error.type == DioExceptionType.connectionTimeout ||
      error.type == DioExceptionType.sendTimeout ||
      error.type == DioExceptionType.receiveTimeout) {
    return '요청 시간이 초과되었습니다. 잠시 후 다시 시도해주세요.';
  }

  if (error.type == DioExceptionType.connectionError) {
    return '서버에 연결할 수 없습니다. 네트워크 상태 또는 서버 실행 상태를 확인해주세요.';
  }

  final data = error.response?.data;
  if (data is Map) {
    final message = data['message'];
    if (message is String && message.isNotEmpty) return message;
    final errorMsg = data['error'];
    if (errorMsg is String && errorMsg.isNotEmpty) return errorMsg;
  }

  final status = error.response?.statusCode;
  if (status != null && status >= 500) {
    return '서버 오류가 발생했습니다. 잠시 후 다시 시도해주세요.';
  }

  return error.message ?? '요청 처리 중 오류가 발생했습니다.';
}

typedef AuthTokenProviderGetter = AuthTokenProvider Function();

void setupAuthInterceptor(
  Dio dio, {
  required AuthTokenProviderGetter tokenProvider,
  required void Function(String message) showError,
}) {
  var isRefreshing = false;
  final pendingQueue = <Completer<String?>>[];

  void resolveQueue(String? token) {
    for (final completer in pendingQueue) {
      if (!completer.isCompleted) completer.complete(token);
    }
    pendingQueue.clear();
  }

  void rejectQueue(Object error) {
    for (final completer in pendingQueue) {
      if (!completer.isCompleted) completer.completeError(error);
    }
    pendingQueue.clear();
  }

  Future<bool> performRefresh() async {
    if (isRefreshing) {
      final completer = Completer<String?>();
      pendingQueue.add(completer);
      try {
        final token = await completer.future;
        return token != null;
      } catch (_) {
        return false;
      }
    }

    isRefreshing = true;
    try {
      final ok = await tokenProvider().refreshToken();
      if (ok) {
        resolveQueue(tokenProvider().tokenResponse?.accessToken);
      } else {
        rejectQueue(StateError('Token refresh failed'));
      }
      return ok;
    } catch (error) {
      rejectQueue(error);
      rethrow;
    } finally {
      isRefreshing = false;
    }
  }

  void emitRedirectToLogin() {
    tokenProvider().clearTokenResponse();
    AuthEventBus.instance.emit(AuthEvent.redirectToLogin);
  }

  dio.interceptors.add(
    InterceptorsWrapper(
      onRequest: (options, handler) async {
        if (_shouldShowLoading(options)) LoadingBus.start();

        if (_isPublicRequest(options.path)) {
          handler.next(options);
          return;
        }

        if (!tokenProvider().isInitialized) {
          await tokenProvider().initializeAuth();
        }

        final token = tokenProvider().tokenResponse;
        final issuedAt = tokenProvider().tokenIssuedAt;

        if (token == null || token.accessToken.isEmpty) {
          handler.next(options);
          return;
        }

        final isLogout = _isLogoutRequest(options.path);

        if (!isLogout && isTokenExpired(token, issuedAt)) {
          if (_shouldShowLoading(options)) LoadingBus.stop();
          emitRedirectToLogin();
          handler.reject(
            DioException(
              requestOptions: options,
              error: 'Token expired',
              type: DioExceptionType.cancel,
            ),
          );
          return;
        }

        if (!isLogout && isTokenExpiringSoon(token, issuedAt)) {
          try {
            final ok = await performRefresh();
            if (!ok) {
              if (_shouldShowLoading(options)) LoadingBus.stop();
              emitRedirectToLogin();
              handler.reject(
                DioException(
                  requestOptions: options,
                  error: 'Token refresh failed',
                  type: DioExceptionType.cancel,
                ),
              );
              return;
            }
          } catch (error) {
            if (_shouldShowLoading(options)) LoadingBus.stop();
            emitRedirectToLogin();
            handler.reject(
              DioException(
                requestOptions: options,
                error: error,
                type: DioExceptionType.cancel,
              ),
            );
            return;
          }
        }

        final accessToken = tokenProvider().tokenResponse?.accessToken;
        if (accessToken != null && accessToken.isNotEmpty) {
          options.headers[authHeader] = '$authType$accessToken';
        }
        handler.next(options);
      },
      onResponse: (response, handler) {
        if (_shouldShowLoading(response.requestOptions)) LoadingBus.stop();
        handler.next(response);
      },
      onError: (error, handler) async {
        final request = error.requestOptions;
        if (_shouldShowLoading(request)) LoadingBus.stop();

        final status = error.response?.statusCode;
        final isAuthRequest =
            _isPublicRequest(request.path) ||
            _isLogoutRequest(request.path) ||
            request.path.contains('/auth/refresh');
        final hasRetried = request.extra['_retry'] == true;
        final shouldHandle401 = status == 401 && !hasRetried && !isAuthRequest;

        if (!shouldHandle401) {
          if (_shouldShowErrorPopup(request) && !isAuthRequest) {
            showError(_resolveErrorMessage(error));
          }
          handler.next(error);
          return;
        }

        request.extra['_retry'] = true;

        try {
          final ok = await performRefresh();
          if (!ok) {
            if (_shouldShowErrorPopup(request)) {
              showError(_resolveErrorMessage(error));
            }
            emitRedirectToLogin();
            handler.next(error);
            return;
          }

          final newToken = tokenProvider().tokenResponse?.accessToken;
          if (newToken != null && newToken.isNotEmpty) {
            request.headers[authHeader] = '$authType$newToken';
          }

          final retryResponse = await dio.fetch<dynamic>(request);
          handler.resolve(retryResponse);
        } catch (refreshError) {
          if (_shouldShowErrorPopup(request)) {
            showError(_resolveErrorMessage(error));
          }
          emitRedirectToLogin();
          handler.next(error);
        }
      },
    ),
  );
}
